ADR/0035-artifact-streams.md

```diff
@@ -0,0 +1,142 @@
+# 35. Artifact Streams
+
+Date started: 2024-02-05
+
+## Status
+
+Proposed
+
+Relates to:
+
+<!-- ... a lot ... -->
+
+Supercedes:
+
+<!-- ... a lot ... -->
+
+## Authors
+
+- Andrew M
+
+## Context
+
+Since the beginning of AppStudio, we have had a dependency on the [Component] and [Application] model that
+originated in the Hybrid Application Service. A [Component] is a single containerized artifact that has
+information on how to build and deploy recorded in a linked [Devfile]. Multiple [Components] can be grouped
+together into a single [Application] to indicate that the should be deployed together via GitOps.
+
+Along with the [gitops-service] which was decided to stop deploying in [ADR 32. Decoupling Deployment],
+the controllers' APIs and the UI for AppStudio have been tightly coupled with the concept of [Application]s
+and [Component]s. For example, 
+
+- PipelineRuns triggered by PipelinesAsCode produced a single image that is tied to a specific [Component]
+  which is assumed to be an OCI artifact.
+- The AppStudio UI (HAC) display PipelineRuns for all [Component]s in a single [Application] together.
+- [Component]s are tied to artifacts produced from a single a git repository and branch.
+
+This model presented multiple challenges including:
+
+- If a user wants to build a new branch of an already onboarded git repository, they would need to create
+  a new [Component] (with a workspace-unique name).
+- [Component]s can be members of only one [Application].
+- As the number of related [Component]s increases, it becomes much harder to visualize all of the currently
+  or previously running PipelineRuns since all [Component]s are included together. This has resulted in users
+  creating multiple [Application]s that are related to each other just to resolve the visual representation
+  challenges.
+- Since a PipelineRun produces a single [Component], it becomes harder to support build strategies that produce
+  multiple artifacts, i.e [ko].
+- [Component]s are only defined in the context of an [Application] and cannot be moved. When a user wants to
+  start building a Component, they need to either add it to a current [Application] and have its builds affecting
+  the IntegrationTestScenarios results of the generated Snapshots or the component needs to be created in its
+  own temporary   Application only to be deleted and recreated in the appropriate one once the builds are
+  properly functioning.
+
+## Decision
+
+We are going to abstract the concepts of the built artifacts and their relationships to each other outside of
+any other controller within AppStudio. If a controller wants to maintain complete separation from others (i.e.
+in order to be fully deployable on its own), it may use an adapter pattern to convert this abstraction to its
+own relevant internal data model.
+
+### ArtifactConfig
```

Comment:
It would be great to mention which CRs should be created by the users and which CRs are created by controllers as a result of a build (for example). For machine generated CRs, I think we should have a plan on how we offload those from the cluster to an external database (like we do for pipeline runs). This is needed in order to keep the history without overloading etcd and will also make migration of tenants between clusters easier.

Reply:
I agree that we should offload the information from the cluster so that we don't continue to hit the problems that we encounter today. A generic tool like [KubeArchive](https://github.com/adambkaplan/kubearchive) would be critical as we will be producing multiple CRs for each artifact produced by the system.

Reply:
I is on our roadmap to maintain[KubeArchive](https://github.com/adambkaplan/kubearchive) ?

Reply:
@gbenhaim the agogos team is considering adopting it. If they do, then the answer to your question is yes.

Note, the kubearchive repo moved to https://github.com/kubearchive/dynowatch

Reply:
Silly question, given that these are going to need to be stored in some kind of off cluster database anyway, did anyone consider moving this data model off cluster into a RDBMS + REST service? If controllers need information about an object they can just query the REST API directly.

This means that current and historic data would all be in the same system, and it would reduce the load on etcd. 

There may be compelling reasons to keep the current on-cluster data storage, but if we are making a big change to the architecture it might be worth considering completely different approaches.

Reply:
I totally agree with @stuartwdouglas. We must consider RDBMS + REST service.

Reply:
I don't think that we have had an ADR behind our decision to use CRs as a data model _and_ API. I assume that this was done back when the architecture was dependent on KCP as these resources would enable data to be shared and referenced regardless of where it was generated/used.

@ralphbean and I have had discussions in the past about whether we should use resources or events as interface models. This discussion falls in a similar vein.